includes/classes/Indexable/Comment/Comment.php

```diff
@@ -0,0 +1,1053 @@
+						],
+					],
+					[
+						'terms' => [
+							'user_id' => array_values( array_map( 'absint', $unapproved_ids ) ),
+						],
+					],
+					[
+						'terms' => [
+							'comment_author_email.raw' => array_values( $unapproved_emails ),
+						],
+					],
+				];
+			} else {
+				$filter['bool']['must'][] = [
+					$terms_map_name => [
+						'comment_approved' => $comment_stati,
+					],
+				];
+			}
+
+			$use_filters = true;
+		}
+
+		/**
+		 * Support `type` query var
+		 */
+		if ( ! empty( $query_vars['type'] ) ) {
+			$types = (array) ( is_string( $query_vars['type'] ) ? explode( ',', $query_vars['type'] ) : $query_vars['type'] );
+			$types = array_map( 'trim', $types );
+
+			$terms_map_name = 'terms';
+
+			if ( count( $types ) < 2 ) {
+				$terms_map_name = 'term';
+				$types          = $types[0];
+			}
+
+			$filter['bool']['must'][] = [
+				$terms_map_name => [
+					'comment_type.raw' => $types,
+				],
+			];
+
+			$use_filters = true;
+		}
+
+		/**
+		 * Support `type__in` query var
+		 */
+		if ( ! empty( $query_vars['type__in'] ) ) {
+			$filter['bool']['must'][]['bool']['must'] = [
+				'terms' => [
+					'comment_type.raw' => array_values( (array) $query_vars['type__in'] ),
+				],
+			];
+
+			$use_filters = true;
+		}
+
+		/**
+		 * Support `type__not_in` query var
+		 */
+		if ( ! empty( $query_vars['type__not_in'] ) ) {
+			$filter['bool']['must'][]['bool']['must_not'] = [
+				'terms' => [
+					'comment_type.raw' => array_values( (array) $query_vars['type__not_in'] ),
+				],
+			];
+
+			$use_filters = true;
+		}
+
+		if ( $use_filters ) {
+			$formatted_args['post_filter'] = $filter;
+		}
+
+		return apply_filters( 'ep_comment_formatted_args', $formatted_args, $query_vars );
+	}
+
+	/**
+	 * Put mapping for comments
+	 *
+	 * @since  3.1
+	 * @return boolean
+	 */
+	public function put_mapping() {
+		$es_version = Elasticsearch::factory()->get_elasticsearch_version();
+
+		if ( empty( $es_version ) ) {
+			$es_version = apply_filters( 'ep_fallback_elasticsearch_version', '2.0' );
+		}
+
+		$mapping_file = 'initial.php';
+
+		if ( version_compare( $es_version, '5.0', '<' ) ) {
+			$mapping_file = 'pre-5-0.php';
+		} elseif ( version_compare( $es_version, '7.0', '>=' ) ) {
+			$mapping_file = '7-0.php';
+		}
+
+		$mapping = require apply_filters( 'ep_comment_mapping_file', __DIR__ . '/../../../mappings/comment/' . $mapping_file );
+
+		$mapping = apply_filters( 'ep_comment_mapping', $mapping );
+
+		return Elasticsearch::factory()->put_mapping( $this->get_index_name(), $mapping );
+	}
+
+	/**
+	 * Query DB for comments
+	 *
+	 * @param  array $args Query arguments
+	 * @since  3.1
+	 * @return array
+	 */
+	public function query_db( $args ) {
+		$all_query = new WP_Comment_Query(
+			[
+				'count' => true,
+			]
+		);
+
+		$defaults = [
+			'number'  => $this->get_bulk_items_per_page(),
+			'offset'  => 0,
+			'orderby' => 'comment_ID',
+			'order'   => 'desc',
+		];
```

Comment:
Need to specify `'type' => 'all'` otherwise EP will only index comments with 'comment' type.
```suggestion
	public function query_db( $args ) {
		$all_query = new WP_Comment_Query(
			[
				'count' => true,
				'type'  => 'all'
			]
		);

		$defaults = [
			'type'    => 'all',
			'number'  => $this->get_bulk_items_per_page(),
			'offset'  => 0,
			'orderby' => 'comment_ID',
			'order'   => 'desc',
		];
```

Reply:
@shadyvb I can't reproduce your issue. What is the ES version used in your testing? Can you give me steps to reproduce it?

Reply:
@dinhtungdu I think it makes sense to explicitly query for `all` comments, instead of allowing the possibility that this might break if the consuming code assumes `''` means only the `comment` type ( which holds true in my case, and a few other plugins that does not support multiple comment types ).

So, it would cover more use cases and be less likely to fail. After all, that's why `'all'` is an option for the `type` argument, so it can be used in situations like this.

Reply:
@shadyvb According to the source of `WP_Comment_Query`, `''` and `'all'` have the same effect, so I don't think we need to specify `all` in this method. See: https://github.com/WordPress/wordpress-develop/blob/master/src/wp-includes/class-wp-comment-query.php#L742-L745

Can you explain to me why other plugins can affect this method, we're querying WordPress comments only?

Reply:
What I was trying to say is that `'all'` should work for ALL cases, but `''` would not, eg: it's the matter of being better, not being right or wrong.

I'd be happy with just adding a filter and letting each consumer adapt to their needs.

```suggestion
	public function query_db( $args ) {
		$all_query_args = apply_filters( 'ep_search_all_comments_args', [
			'count' => true,
		] );
		
		$all_query = new WP_Comment_Query( $all_query_args );

		$defaults = apply_filters( 'ep_search_paginated_comments_args', [
			'number'  => $this->get_bulk_items_per_page(),
			'offset'  => 0,
			'orderby' => 'comment_ID',
			'order'   => 'desc',
		] );
```

@dinhtungdu Does that make more sense ?

Reply:
@shadyvb Thanks for the clarification, I added a new filter to all query as you suggested. I merged some of your suggestions as well. Great work! I really appreciate all the work you have done!